src/transformStateWithClones.js

```diff
@@ -7,7 +7,29 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let stateCopy = { ...state };
+  const stateCopyChanges = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateCopy, action.extraData);
+        break;
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete stateCopy[key];
+        }
+        break;
+      case 'clear':
+        stateCopy = {};
+        break;
+      default:
+        throw new Error(`Unknown action type '${action.type}'`);
+    }
+    stateCopyChanges.push({ ...stateCopy });
+  }
+
+  return stateCopyChanges;
 }
 
 module.exports = transformStateWithClones;
```

Comment:
Consider using a more descriptive name for the array `stateCopyChanges`. The name should reflect that it contains the history of state changes. Something like `stateHistory` or `stateSnapshots` would be more informative.